README.md
# addon-opened-by

An add-on for handling warnings/locks when records are opened by someone else. OpenedBy uses REST-api calls to create OpenedBy records which also is queried upon opening records of a certain type. 

Only works in the desktop client. 

Is cloud compatible.

## Prerequisites
To make OpenedBy work properly the handling of GeneralInspectorHandler and HelpdeskInspectorHandler(or ) need to be written in a certain manner. Use the following example as guide:

GeneralInspectorHandler:
```
Private Sub m_Application_AfterActiveInspectorChanged()
    On Error GoTo ErrorHandler
    If Not m_Application.ActiveInspector Is Nothing Then
        Set m_Inspector = m_Application.ActiveInspector
        
        Select Case m_Inspector.Class.Name
            Case "helpdesk":
                Const tagName As String = "HelpdeskInspector_Listener"
                If Not m_Inspector Is Nothing Then
                    If Not IsObject(m_Inspector.Tag(tagName)) Then
                        Dim helpdeskListener As New HelpdeskInspectorHandler
                        Call helpdeskListener.Connect(m_Inspector, tagName)                        
                    End If
                End If
        End Select
    End If
```

HelpdeskInspectorHandler:
```
(General) (Declarations)
Private m_tagName As String

Private Sub m_Inspector_AfterClose()
    Disconnect
End Sub

Public Sub Connect(limeinspector As Lime.Inspector, tagName As String)
    m_tagName = tagName
    Set m_Inspector = limeinspector
    m_Inspector.Tag(m_tagName) = Me
End Sub

Public Sub Disconnect()
    If Not m_Inspector Is Nothing Then
        m_Inspector.Tag(m_tagName) = Nothing
        Set m_Inspector = Nothing
    End If
End Sub
```
## Installation

Install the package with LIP. A VBA .bas file will be installed as well as a new table called OpenedBy along with some localization records.

Add the following line of code in the ControlsHandler.Class_Initialize of your desire:

```
Call OpenedBy.SetOpenedBy(m_controls.Record.ID, "<tablename>")
```

To delete the OpenedBy post you add the following line of code to the BeforeClose method in the InspectorHandler of your desire and to the BeforeRecordChanged in the ControlsHandler of your desire:

ControlsHandler.m_controls.BeforeRecordChanged:
```
'Opened By
If Not m_inspector Is Nothing Then
    Call OpenedBy.RemoveOpenedBy(m_Controls.Record.ID, m_Controls.Class.Name)
End If

```

InspectorHandler.BeforeClose:
```
Call OpenedBy.RemoveOpenedBy(m_inspector.Record.ID, m_inspector.Class.Name)
```

The following code should be added to the ExplorerHandler.BeforeCommand of your choice (remember to check for the right class if you're using the GeneralExplorerHandler):

```
'OpenedBy --------->
If Command = lkCommandOpen Then        
    If Not ActiveUser Is Nothing Then
        Cancel = OpenedBy.IsOpenedBy(m_explorer.ActiveItem.ID, "<tablename>")
    End If
End If
'<-------- OpenedBy
```

Finally you need to change the url for the REST-api which is defined in the General.Declarations of the OpenedBy.bas in VBA. There you can also choose if you want to block others from opening records that already are open.
```
'Change this if you want Lime to block locked posts
Const bBlockOnOpen As Boolean = False
'Change this according to your api adress
Const sUrlBase As String = "https://<server_name>/<app_name>/api/v1/limeobject/openedby/"
```
You can change the message through the localize posts that are related to the OpenedBy package.

If on premise, to clean up OpenedBy records that have accidentally been left behind (ex. if Lime crashes), add the ```csp_clear_opened_by``` stored procedure under /sql folder, and add a job that runs it in the SQL Server Agent. Schedule it to run nightly. There is an example also under /sql folder. Enter needed changes:

* ```<ENTER NAME OF AGENT JOB HERE>```
* ```<USER RUNNING THE JOB>```
* ```<ENTER DATABASE NAME HERE>```

In Cloud this is not possible so there is a VBA method, CleanUpExpired, which clears OpenedBy records that are older than a day (this is a setting in the OpenedBy module). Add a call to the method in the ThisApplication method which suits best for it.  

The OpenedBy table should also be visible for administrators (it is by default) so that they can remove records manually if needed.




CHANGELOG.md
### [1.0.3](https://github.com/Lundalogik/addon-opened-by/compare/v1.0.2...v1.0.3) (2021-07-01)


### Bug Fixes

* trigger new release ([4b68729](https://github.com/Lundalogik/addon-opened-by/commit/4b68729964d7ee76e1db4e0b3f6ac0a6665985f8))

### [1.0.2](https://github.com/Lundalogik/addon-opened-by/compare/v1.0.1...v1.0.2) (2021-07-01)


### Bug Fixes

* added ci folder ([67f8225](https://github.com/Lundalogik/addon-opened-by/commit/67f822500255ff790bffb25227311d28578b6dd6))

### [1.0.1](https://github.com/Lundalogik/addon-opened-by/compare/v1.0.0...v1.0.1) (2021-07-01)


### Bug Fixes

* to trigger new release ([3b3f55e](https://github.com/Lundalogik/addon-opened-by/commit/3b3f55e57bf942f5308c73a7611f43461af3b88b))

## 1.0.0 (2021-06-28)


### Features

* added code for agent job and important documentation for handling deletion of openedby records ([4d2ccb1](https://github.com/Lundalogik/addon-opened-by/commit/4d2ccb1bf5396b1d2d01d23145d24e4ff1d0aa03))


### Bug Fixes

* changed doc ([6cefda9](https://github.com/Lundalogik/addon-opened-by/commit/6cefda994de33248d6b851173ea6fd238b6d79f3))
* fixed agent job ([ab86482](https://github.com/Lundalogik/addon-opened-by/commit/ab86482c4613597b098a3d0ee0ddf023f97053a0))
* trigger new release ([5a25cd4](https://github.com/Lundalogik/addon-opened-by/commit/5a25cd416d4554c99606ed92ab84faa78453c877))
* trigger new release ([1dca3bd](https://github.com/Lundalogik/addon-opened-by/commit/1dca3bd76c41935733f03c9764d10fe5b9d0856b))
* trigger new release ([96293d7](https://github.com/Lundalogik/addon-opened-by/commit/96293d768eecf1628ca9ecd8b3e07d7642fceaef))


ci/toolbox.py
import os
import click
import sys
import requests
import re
import subprocess


@click.group()
def cli():
    pass


@cli.command()
@click.argument('repository')
@click.argument('version')
def set_env_from_gh_release_api(repository, version):
    """
        Set the following ENV VARIABLES is gh-actions env
        => RELEASE_UPLOAD_URL
    """
    def _get_dispatch_token():
        token = os.environ.get('DISPATCH_TOKEN', None)
        if not token:
            click.echo("DISPATCH_TOKEN token variable is missing")
            sys.exit(1)
        return token

    def _set_env(name, value):
        cmd = 'echo "' + name + '=' + value + '" >> $GITHUB_ENV'
        subprocess.call(cmd, shell=True)

    dispatch_token = _get_dispatch_token()
    headers = {"Authorization": f"token {dispatch_token}"}
    url = f"https://api.github.com/repos/Lundalogik/{repository}/releases/tags/v{version}"  # noqa
    response = requests.get(url, headers=headers)
    response.raise_for_status()

    data = response.json()
    _set_env('RELEASE_UPLOAD_URL', data['upload_url'])


@cli.command()
@click.argument('release_notes')
def update_docs_changelog(release_notes):
    """
        Parsing release notes and update docs/changelog.md file
    """

    def _format_release_version():
        first_line = release_notes.splitlines()[0]

        # Fetch release version from first line in release note
        match = re.search(r"\[([\d.\d.\d]+)\]", first_line)
        if match:
            return f"## v{match.group(1)}\n\n"

        return "## unknown version\n\n"

    def _format_release_date():
        first_line = release_notes.splitlines()[0]

        # Fetch release date from first line in release note
        match = re.search(
            r"([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))",
            first_line)
        if match:
            return f"**Released:** {match.group(1)}"

        return "**Released:** unknown date"

    def _format_release_notes():
        tail = release_notes.split("\n", 1)[1]

        # Strip injected links
        matches = [x.group for x in re.finditer(r"\(\[(.*)\]\((.*)\)\)", tail)]
        for match in matches:
            tail = tail.replace(match(), "")

        # Strip trailing space on each row
        tail = tail.replace(" \n", "\n")

        # Strip trailing new lines at end of string
        tail = tail.rstrip("\n")

        return f"{tail}\n"

    def _format_new_release():
        """
            Desired format =>

                ## VERSION

                **Released:** DATE

                ### Bug Fixes

                * FIX_ONE
                * FIX_TWO

                ### FEATURES

                * FEAT_ONE
                * FEAT_TWO
        """
        return (f"{_format_release_version()}"
                f"{_format_release_date()}"
                f"{_format_release_notes()}")

    with open("docs/changelog.md", "r") as original_file:
        old_releases = ""

        if original_file.read(1):
            original_file.seek(0)

            lines = original_file.readlines()
            if len(lines) > 1:
                lines.pop(0)
                old_releases = ''.join(lines)

    with open('docs/changelog.md', 'w') as modified_file:
        title = "# Changelog\n\n"
        new_release = _format_new_release()

        modified_file.write(title)
        modified_file.write(new_release)
        modified_file.write(old_releases)


if __name__ == '__main__':
    cli()


legacy/lip/vba/AO_OpenedBy.bas
Attribute VB_Name = "AO_OpenedBy"
Option Explicit
'Settings
'Change this if you want Lime to block locked posts
Const bBlockOnOpen As Boolean = False
'Change this according to your api adress
Const sUrlBase As String = "https://<server name>/<database name>/api/v1/limeobject/openedby/"
'Change this
Const iExpireDays As Integer = 1

Public Sub CleanUpExpired()
    ' mwe quick hack to clean up (2019-11-26)
    On Error GoTo errorhandler

    Dim pFilter As New LDE.Filter
    Dim pRecords As New LDE.Records
    Dim pRecord As New LDE.Record
    Dim pBatch As New LDE.Batch
    Set pBatch.Database = Application.Database
    
    Call pFilter.AddCondition("createdtime", lkOpLess, Now - 1)

    Call pRecords.Open(Classes("openedby"), pFilter)
    For Each pRecord In pRecords
        pRecord.delete
        Call pRecord.Update(pBatch)
    Next
    Call pBatch.Execute
    Exit Sub
errorhandler:
End Sub




Function IsOpenedBy(sRecordId As String, sTableName As String) As Boolean
    On Error GoTo errorhandler
    
    Dim sResponse As String
    Dim sMyUrl As String
    Dim oJson As Object
    Dim embedded As Object
    Dim iAnswer As Integer
    Dim sOpenedByUser As String
    Dim bIsOpenedByOther As Boolean
    Dim i As Integer
    Dim bCancelOpen As Boolean
    
    'Use the REST API to see if someone has opened the record already
    sMyUrl = sUrlBase + "?recordid=" + sRecordId + "&tablename=" + sTableName
    sResponse = XmlHttpSend("GET", sMyUrl, Nothing)
    
    'Warning should not be shown
    bCancelOpen = False
    'Warning should be shown
    IsOpenedBy = bCancelOpen
    
    If sResponse <> "" Then
        Set oJson = JSON.parse(sResponse)
        Set embedded = oJson("_embedded")
        
        If embedded("limeobjects").count <> 0 Then
            'If post is opened by another user
            For i = 1 To embedded("limeobjects").count
                If embedded("limeobjects")(i)("iduser") <> CStr(Application.ActiveUser.ID) Then
                    bIsOpenedByOther = True
                    sOpenedByUser = embedded("limeobjects")(i)("name")
                    Exit For
                End If
            Next i
            If bIsOpenedByOther Then
                'If record already is opened then block or allow the iduser to proceed
                If Not bBlockOnOpen Then
                    iAnswer = Lime.MessageBox(Localize.GetText("AO_OpenedBy", "i_openedbymessage"), VBA.vbYesNo + VBA.vbQuestion + vbDefaultButton2, sOpenedByUser)
                    If iAnswer = vbYes Then
                        bCancelOpen = False
                        IsOpenedBy = bCancelOpen
                    Else
                        bCancelOpen = True
                        IsOpenedBy = bCancelOpen
                    End If
                Else
                    Call Lime.MessageBox(Localize.GetText("AO_OpenedBy", "i_blockedopenedby"), vbOKOnly + vbExclamation, sOpenedByUser)
                    bCancelOpen = True
                    IsOpenedBy = bCancelOpen
                End If
            End If
        End If
    End If

    Exit Function
errorhandler:
    IsOpenedBy = True
    Call LC_UI.ShowError("AO_OpenedBy.IsOpenedBy")
End Function

Sub SetOpenedBy(sRecordId As String, sTableName As String)
    On Error GoTo errorhandler
    
    Dim sResponse As String
    Dim oPayloadBody As New Scripting.Dictionary
    
    oPayloadBody.Add "iduser", ActiveUser.ID
    oPayloadBody.Add "recordid", sRecordId
    oPayloadBody.Add "tablename", sTableName
    oPayloadBody.Add "name", ActiveUser.Name
        
    sResponse = XmlHttpSend("POST", sUrlBase, oPayloadBody)
    
    Exit Sub
errorhandler:
    Call LC_UI.ShowError("AO_OpenedBy.SetAsOpenedBy")
End Sub

Sub RemoveOpenedBy(sRecordId As String, sTableName As String)
    On Error GoTo errorhandler
    
    'Settings
    Dim sResponse As String
    Dim geturl As String
    Dim deleteurl As String
    Dim ID As String
    Dim limeobjects As Collection
    Dim embedded As Object
    Dim oJson As Object
    Dim i As Integer
    
    'REST API GET idopenedby from tablename
    geturl = sUrlBase & "?recordid=" & sRecordId & "&iduser=" & ActiveUser.ID & "&tablename=" & sTableName
    sResponse = XmlHttpSend("GET", geturl, Nothing)

    'REST API DELETE if idopenedby found
    If sResponse <> "" Then
        Set oJson = JSON.parse(sResponse)
        Set embedded = oJson("_embedded")
        If embedded("limeobjects").count > 0 Then
            For i = 1 To embedded("limeobjects").count
                Set limeobjects = embedded("limeobjects")
                ID = CStr(limeobjects(i)("_id"))
            
                deleteurl = sUrlBase & ID & "/"
        
                sResponse = AO_OpenedBy.XmlHttpSend("DELETE", deleteurl, Nothing)
            Next
        End If
    End If
    
    Exit Sub
errorhandler:
    Call LC_UI.ShowError("AO_OpenedBy.RemoveOpenedBy")
End Sub

Function XmlHttpSend(sMethod As String, sUrl As String, payload As Scripting.Dictionary) As String
    On Error GoTo errorhandler
    
    Dim xmlhttp As New MSXML2.XMLHTTP60
    
    xmlhttp.Open sMethod, sUrl, False
    xmlhttp.setRequestHeader "Content-Type", "application/json"
    xmlhttp.setRequestHeader "sessionid", Application.Database.SessionID
    
    If payload Is Nothing Then
        xmlhttp.Send
    Else
        xmlhttp.Send JsonConverter.ConvertToJson(payload)
    End If
    
    If xmlhttp.Status > 199 And xmlhttp.Status < 205 Then
        XmlHttpSend = xmlhttp.responseText
    Else
        Lime.MessageBox ("API call returned " & xmlhttp.Status & " " & xmlhttp.statusText), vbOKOnly + vbExclamation
        XmlHttpSend = ""
    End If
    
    Exit Function
errorhandler:
    Call LC_UI.ShowError("AO_OpenedBy.XmlHttpSend")
End Function


legacy/vba/OpenedBy.bas
Attribute VB_Name = "OpenedBy"
Option Explicit
'Settings
'Change this if you want Lime to block locked posts
Const bBlockOnOpen As Boolean = False
'Change this according to your api adress
Const sUrlBase As String = "https://<server name>/<database name>/api/v1/limeobject/openedby/"
'Change this
Const iExpireDays As Integer = 1

Public Sub CleanUpExpired()
    ' mwe quick hack to clean up (2019-11-26)
    On Error GoTo errorhandler

    Dim pFilter As New LDE.Filter
    Dim pRecords As New LDE.Records
    Dim pRecord As New LDE.Record
    Dim pBatch As New LDE.Batch
    Set pBatch.Database = Application.Database
    
    Call pFilter.AddCondition("createdtime", lkOpLess, Now - 1)

    Call pRecords.Open(Classes("openedby"), pFilter)
    For Each pRecord In pRecords
        pRecord.delete
        Call pRecord.Update(pBatch)
    Next
    Call pBatch.Execute
    Exit Sub
errorhandler:
End Sub




Function IsOpenedBy(sRecordId As String, sTableName As String) As Boolean
    On Error GoTo errorhandler
    
    Dim sResponse As String
    Dim sMyUrl As String
    Dim oJson As Object
    Dim embedded As Object
    Dim iAnswer As Integer
    Dim sOpenedByUser As String
    Dim bIsOpenedByOther As Boolean
    Dim i As Integer
    Dim bCancelOpen As Boolean
    
    'Use the REST API to see if someone has opened the record already
    sMyUrl = sUrlBase + "?recordid=" + sRecordId + "&tablename=" + sTableName
    sResponse = XmlHttpSend("GET", sMyUrl, Nothing)
    
    'Warning should not be shown
    bCancelOpen = False
    'Warning should be shown
    IsOpenedBy = bCancelOpen
    
    If sResponse <> "" Then
        Set oJson = JSON.parse(sResponse)
        Set embedded = oJson("_embedded")
        
        If embedded("limeobjects").count <> 0 Then
            'If post is opened by another user
            For i = 1 To embedded("limeobjects").count
                If embedded("limeobjects")(i)("iduser") <> CStr(Application.ActiveUser.ID) Then
                    bIsOpenedByOther = True
                    sOpenedByUser = embedded("limeobjects")(i)("name")
                    Exit For
                End If
            Next i
            If bIsOpenedByOther Then
                'If record already is opened then block or allow the iduser to proceed
                If Not bBlockOnOpen Then
                    iAnswer = Lime.MessageBox(Localize.GetText("OpenedBy", "i_openedbymessage"), VBA.vbYesNo + VBA.vbQuestion + vbDefaultButton2, sOpenedByUser)
                    If iAnswer = vbYes Then
                        bCancelOpen = False
                        IsOpenedBy = bCancelOpen
                    Else
                        bCancelOpen = True
                        IsOpenedBy = bCancelOpen
                    End If
                Else
                    Call Lime.MessageBox(Localize.GetText("OpenedBy", "i_blockedopenedby"), vbOKOnly + vbExclamation, sOpenedByUser)
                    bCancelOpen = True
                    IsOpenedBy = bCancelOpen
                End If
            End If
        End If
    End If

    Exit Function
errorhandler:
    IsOpenedBy = True
    Call LC_UI.ShowError("OpenedBy.IsOpenedBy")
End Function

Sub SetOpenedBy(sRecordId As String, sTableName As String)
    On Error GoTo errorhandler
    
    Dim sResponse As String
    Dim oPayloadBody As New Scripting.Dictionary
    
    oPayloadBody.Add "iduser", ActiveUser.ID
    oPayloadBody.Add "recordid", sRecordId
    oPayloadBody.Add "tablename", sTableName
    oPayloadBody.Add "name", ActiveUser.Name
        
    sResponse = XmlHttpSend("POST", sUrlBase, oPayloadBody)
    
    Exit Sub
errorhandler:
    Call LC_UI.ShowError("OpenedBy.SetAsOpenedBy")
End Sub

Sub RemoveOpenedBy(sRecordId As String, sTableName As String)
    On Error GoTo errorhandler
    
    'Settings
    Dim sResponse As String
    Dim geturl As String
    Dim deleteurl As String
    Dim ID As String
    Dim limeobjects As Collection
    Dim embedded As Object
    Dim oJson As Object
    Dim i As Integer
    
    'REST API GET idopenedby from tablename
    geturl = sUrlBase & "?recordid=" & sRecordId & "&iduser=" & ActiveUser.ID & "&tablename=" & sTableName
    sResponse = XmlHttpSend("GET", geturl, Nothing)

    'REST API DELETE if idopenedby found
    If sResponse <> "" Then
        Set oJson = JSON.parse(sResponse)
        Set embedded = oJson("_embedded")
        If embedded("limeobjects").count > 0 Then
            For i = 1 To embedded("limeobjects").count
                Set limeobjects = embedded("limeobjects")
                ID = CStr(limeobjects(i)("_id"))
            
                deleteurl = sUrlBase & ID & "/"
        
                sResponse = OpenedBy.XmlHttpSend("DELETE", deleteurl, Nothing)
            Next
        End If
    End If
    
    Exit Sub
errorhandler:
    Call LC_UI.ShowError("OpenedBy.RemoveOpenedBy")
End Sub

Function XmlHttpSend(sMethod As String, sUrl As String, payload As Scripting.Dictionary) As String
    On Error GoTo errorhandler
    
    Dim xmlhttp As New MSXML2.XMLHTTP60
    
    xmlhttp.Open sMethod, sUrl, False
    xmlhttp.setRequestHeader "Content-Type", "application/json"
    xmlhttp.setRequestHeader "sessionid", Application.Database.SessionID
    
    If payload Is Nothing Then
        xmlhttp.Send
    Else
        xmlhttp.Send JsonConverter.ConvertToJson(payload)
    End If
    
    If xmlhttp.Status > 199 And xmlhttp.Status < 205 Then
        XmlHttpSend = xmlhttp.responseText
    Else
        Lime.MessageBox ("API call returned " & xmlhttp.Status & " " & xmlhttp.statusText), vbOKOnly + vbExclamation
        XmlHttpSend = ""
    End If
    
    Exit Function
errorhandler:
    Call LC_UI.ShowError("OpenedBy.XmlHttpSend")
End Function
